csrc/val_graph.cpp

```diff
@@ -335,10 +335,13 @@ void ValGraph::mapVals(Val* val0, Val* val1) {
   const ValGroup orig_val_group0 = toGroup(val0);
   const ValGroup orig_val_group1 = toGroup(val1);
 
-  const ExprGroups& orig_defs0 = getDefinitions(orig_val_group0);
-  const ExprGroups& orig_defs1 = getDefinitions(orig_val_group1);
-  const ExprGroups& orig_uses0 = getUses(orig_val_group0);
-  const ExprGroups& orig_uses1 = getUses(orig_val_group1);
+  // Note that getDefinitions and getUses return references, which
+  // will be invalidated once unique_definitions_ and unique_uses_ are
+  // updated
+  const ExprGroups orig_defs0 = getDefinitions(orig_val_group0);
+  const ExprGroups orig_defs1 = getDefinitions(orig_val_group1);
```

Comment:
I believe you put this here for a reason, but I wonder what the effect would be if we put these lines down below after the "uses" loop, so that the "defs" loop would use updated definitions instead of the outdated ones.

Reply:
I think we could reorder them as long as each refers to the correct mapping. The interface is not the most robust design as there are a couple of things that can be easily missed. For example, `disjoint_vals_.mapEntries(val0, val1)` invalidates the references returned by `toGroup` above (that's why they need to copy the returned references). Also, the update of `unique_definitions_` and `unique_uses_` invalidates the references returned by `getDefinitions` and `getUses`. I already had nasty bugs a couple of times 😃